src/metkit/codes/api/CodesHandleRef.h

```diff
@@ -0,0 +1,121 @@
+public:  // methods
+
+    CodesHandleRef(CodesHandlePtr*);
+
+    CodesHandleRef(const CodesHandleRef&)            = default;
+    CodesHandleRef(CodesHandleRef&&)                 = default;
+    CodesHandleRef& operator=(const CodesHandleRef&) = default;
+    CodesHandleRef& operator=(CodesHandleRef&&)      = default;
+
+
+    // Overriding virtual methods
+
+    /// Returns size of the message (number of bytes)
+    size_t messageSize() const override;
+
+    /// Check if a key is defined
+    bool isDefined(const std::string&) const override;
+
+    /// Check if a key is missing
+    bool isMissing(const std::string&) const override;
+
+    /// Check if a key is defined and not missing
+    bool has(const std::string&) const override;
+
+    /// Set a key to its missing value
+    void setMissing(const std::string& key) override;
+
+    /// Set scalars
+    void set(const std::string& key, const std::string& value) override;
+    void set(const std::string& key, double value) override;
+    void set(const std::string& key, long value) override;
+
+    /// Set arrays
+    void set(const std::string& key, Span<const std::string> value) override;  /// set string array
+    void set(const std::string& key, Span<const char*>) override;              /// set string array
+    void set(const std::string& key, Span<const double>) override;
+    void set(const std::string& key, Span<const float> value) override;
+    void set(const std::string& key, Span<const long> value) override;
+    void set(const std::string& key, Span<const unsigned char> value) override;  /// set bytes
+    void forceSet(const std::string& key, Span<const double> value) override;
+    void forceSet(const std::string& key, Span<const float> value) override;
+
+    /// Returns size of the value contained for a given key (can be used to determine if an array is contained)
+    size_t getSize(const std::string& key) const override;
+
+    /// Get the value of the key
+    Value get(const std::string& key) const override;
+
+    /// Get the type of the key
+    NativeType getType(const std::string& key) const override;
+
+    /// Explicit getters
+    long getLong(const std::string& key) const override;
+    double getDouble(const std::string& key) const override;
+    std::string getString(const std::string& key) const override;
+
+    std::vector<long> getLongArray(const std::string& key) const override;
+    std::vector<double> getDoubleArray(const std::string& key) const override;
+    std::vector<float> getFloatArray(const std::string& key) const override;
+    std::vector<std::string> getStringArray(const std::string& key) const override;
+
+    std::vector<unsigned char> getBytes(const std::string& key) const override;
+
+    /// Cloning the whole handle. Expected to be wrapped by the user explicitly
+    [[nodiscard]]
+    CodesHandlePtr* clone() const override;
+
+    /// Copy the message into a new allocated buffer
+    ByteArray copy() const override;
```

Comment:
I think you should implement this as a copyInto, where the caller supplies you with a buffer of his choice, this allows the caller full control over his memory.
This signature would become `void copyInto(uint8_t* buffer)`

Also I think you could then drop the `ByteArray` type, I don't see any other user.

Reply:
Done.